lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -2143,51 +2143,261 @@ namespace Js
     }
 #endif
 
-    /*
-    * This function has the following options:
-    *  - Format as Percent.
-    *  - Format as Number.
-    *  - If significant digits are present, format using the significant digts;
-    *  - Otherwise format using minimumFractionDigits, maximumFractionDigits, minimumIntegerDigits
-    */
+#ifdef INTL_ICU
+    // ICU reports back number format parts differently than date format parts.
+    // Specifically, parts of numbers can have sub-parts, but those sub-parts can't themselves have sub-parts.
+    // Thus, it returns a linear representation of a tree, where the implicit root node represents the entire
+    // formatted string and each child represents a specific part of the string. Per ICU documentation, parts
+    // will *never* overlapped -- any given part can only be a strict subset of exactly one part, a strict
+    // superset of one or more parts, or no relationships at all.
+    // The tree we want to form looks as follows:
+    //
+    // [1] [2] [    3    ] [4] [    5    ]
+    //              |             |   |
+    //             [6]           [7] [8]
+    //
+    // We want to convert the above tree back into an array of parts that looks as follows:
+    //
+    // [1], [2], [3.1], [6], [3.2], [4], [5.1], [7], [5.2], [8], [5.3]
+    //
+    // In more concrete words, we want to assign the entire formatted string to some number of parts,
+    // and each character should get the most specific part possible. We can accomplish this by
+    // keeping an auxiliary array with the same size as the formatted string that, at each index,
+    // holds the part for that character. For the following call pattern:
```

Comment:
I find this explanation difficult to follow. The example that you give below, does that relate to the tree/example above? The example above has 11 entries, while below you refer to `[0,9)` which only has 10.

I'm also not clear on how the array of `0`s below ends up turning into 1,000,000

Reply:
Yeah it took me a while to get to this data structure so I think those are from two different nights. I can try clarify the comment a bit. The goal of the structure is basically that ICU reports back a tree of parts where each node in the tree has a type and a width, and there must be at least one node corresponding to each character in the tree (nodes can be wider than one character). Nodes can have children, and the parent-child relationship is that child node represents a more specific type for a given character range than the parent. Since ICU doesn't ever report "literal" parts of strings (like spaces or other extra characters), the root node in the tree will always be the entire width of the string with the type UnsetField. Then, for a string like "US$ 1,000", there will be two child nodes, one of type currency with width [0, 3) and one of type integer with width [4, 9). The integer node will have a child of type group and width [6, 7). So the most specific type for characters 0 to 3 is currency, 3 to 4 is unset (literal), 4 to 5 is integer, 5 to 6 is group, and 6 to 9 is integer.

I thought about this a bunch and I didn't like the idea of traversing an actual tree structure to get that information because it felt awkward to encode the "width" of nodes with specific meaning. So, I came up with the array structure where basically when we are told a part exists from position x to y, we can figure out what type used to apply to that span and update that section of the array with the new type. We skip over sections of the span [x, y) that have a type that doesn't match the start and end because that means we have already gotten a more specific part for that sub-span (for instance if we got a grouping separator before it's parent integer)

Does that make more sense? If not I can try to figure out a clearer explanation in person on Monday.

, the left-most node would contain the part corresponding to some substring starting at the first character of the overall string, and the right-most node would contain the part corresponding to some substring ending at the end of the overall string. There is probably an actual data structure word for this, but maybe a more specific graph for the case of 1,000,000 is:

```
[                          with